supertokens_python/recipe/session/recipe_implementation.py

```diff
@@ -46,10 +48,107 @@
 if TYPE_CHECKING:
     from typing import List, Union
     from supertokens_python import AppInfo
-    from supertokens_python.querier import Querier
 
-from .constants import JWKCacheMaxAgeInMs, JWKRequestCooldownInMs
 from .interfaces import SessionContainer
+from supertokens_python.querier import Querier
+
+
+class GetJWKSResult:
+    def __init__(self, keys: List[PyJWK], last_fetched: int):
+        self.keys = keys
+        self.last_fetched = last_fetched
+
+
+from typing_extensions import TypedDict
+import threading
+from os import environ
+
+
+class JWKSConfigType(TypedDict):
+    cache_max_age: int
+    refresh_rate_limit: int
+
+
+JWKSConfig: JWKSConfigType = {
+    "cache_max_age": 6000,
+    "refresh_rate_limit": 500,
+}
+
+jwks_cache: Optional[GetJWKSResult] = None
+mutex = threading.RLock()
+
+
+def get_jwks_cache():
+    return jwks_cache
+
+
+def reset_jwks_cache():
+    global jwks_cache
+    jwks_cache = None
+
+
+urls_attempted_for_jwks_fetch: List[str] = []
+
+
+def get_jwks_from_cache_if_present() -> Optional[GetJWKSResult]:
+    with mutex:  # does mutex.acquire() and mutex.release() automatically
+        if jwks_cache is not None:
+            # This means that we have valid JWKs for the given core path
+            # We check if we need to refresh before returning
+
+            # This means that the value in cache is not expired, in this case we return the cached value
+            # Note that this also means that the SDK will not try to query any other Core (if there are multiple)
+            # if it has a valid cache entry from one of the core URLs. It will only attempt to fetch
+            # from the cores again after the entry in the cache is expired
+            now = get_timestamp_ms()
+            if (now - jwks_cache.last_fetched) < JWKSConfig["cache_max_age"]:
+                if environ.get("SUPERTOKENS_ENV") == "testing":
+                    log_debug_message("Returning JWKS from cache")
+                return jwks_cache
+    return None
+
+
+def get_jwks() -> Optional[List[PyJWK]]:
+    global jwks_cache
+
+    core_paths = Querier.get_instance().get_all_core_urls_for_path(
+        "./.well-known/jwks.json"
+    )
+
+    if len(core_paths) == 0:
+        raise Exception(
+            "No SuperTokens core available to query. Please pass supertokens > connectionURI to the init function, or override all the functions of the recipe you are using."
+        )
+
+    results_from_cache = get_jwks_from_cache_if_present()
+    if results_from_cache is not None:
+        return results_from_cache.keys
+
+    with mutex:  # automatically exists on unhandled errors
+        for path in core_paths:
+            if environ.get("SUPERTOKENS_ENV") == "testing":
+                log_debug_message("Attempting to fetch JWKS from path: %s", path)
+
+            # Fetch JWKS again if the kid in the header of the JWT does not match any in
```

Comment:
The comment is incomplete, also does not make sense in the context of the code snippet

Reply:
Creating client now and using `get_matching_key_from_jwt` which fetches again on unknown rid.

Reply:
This is done imo.